File: CppExercise/src/lib/dynamics.cpp
#include <rlab/dynamics.h>

using namespace casadi;

// Constructor implementation
Dynamics::Dynamics(bool is_implicit, bool actuator_dynamics) :
    is_implicit(is_implicit), actuator_dynamics(actuator_dynamics) {

    X = SX::vertcat({
        SX::sym("x"), 
        SX::sym("z"),
        SX::sym("theta"),
        SX::sym("vx"),
        SX::sym("vz"),
        SX::sym("wy")
    });

    n_X = X.size1();

    U = SX::vertcat({
        SX::sym("T"), 
        SX::sym("gimbal")
    });

    n_U = U.size1();
    
    dt = SX::sym("dt");

    if (actuator_dynamics) {
        X = SX::vertcat({
            X, 
            SX::sym("gim_real"), 
            SX::sym("gim_dot")
        });
        n_X = X.size1();

        setupDynamicsWithTVC();
    } else {
        setupDynamicsWithoutTVC();
    }
    setupIntegrator();
}

void Dynamics::setupDynamicsWithTVC() {
    // Define the dynamics with TVC dynamics
    std::cout << "Dynamics with TVC dynamics" << std::endl;
    // State variables
    SX theta    = X(2);
    SX vx       = X(3);
    SX vz       = X(4);
    SX wy       = X(5);
    SX gim_real = X(6);
    SX gim_dot  = X(7);

    // Control variables
    SX T    = U(0); // Thrust
    SX gim  = U(1); // Gimbal angle (input reference)

    // Thrust in body frame
    SX Tz_body = T * cos(gim_real);  // Vertical thrust
    SX Tx_body = T * sin(gim_real);  // Horizontal thrust
    SX T_body  = SX::vertcat({Tx_body, Tz_body});

    // Rotation matrix
    SX Rot = SX::vertcat({
        SX::horzcat({cos(theta), -sin(theta)}),
        SX::horzcat({sin(theta),  cos(theta)})
    });

    // Thrust in inertial frame
    SX T_inert = mtimes(Rot, T_body);
    SX Tx_inert = T_inert(0);
    SX Tz_inert = T_inert(1);

    // Torque from gimbal
    SX tau = T * sin(gim_real) * E_z;

    // Dynamics
    SX x_dot      = vx;
    SX z_dot      = vz;
    SX theta_dot  = wy;
    SX vx_dot     = Tx_inert / m;
    SX vz_dot     = Tz_inert / m - g;
    SX wy_dot     = tau / J_y;
    SX gim_ddot   = -2 * zeta * w_n * gim_dot - w_n * w_n * gim_real + w_n * w_n * gim;

    // State derivative vector
    X_dot = SX::vertcat({
        x_dot,
        z_dot,
        theta_dot,
        vx_dot,
        vz_dot,
        wy_dot,
        gim_dot,
        gim_ddot
    });

    // CasADi function
    f = Function("f", {X, U}, {X_dot}, {"X", "U"}, {"X_dot"});
}

void Dynamics::setupDynamicsWithoutTVC() {
    // Define the dynamics without TVC dynamics

    // State variables
    SX theta    = X(2);
    SX vx       = X(3);
    SX vz       = X(4);
    SX wy       = X(5);

    // Control variables
    SX T    = U(0); // Thrust
    SX gim  = U(1); // Gimbal angle (input reference)

    // Thrust in body frame
    SX Tz_body = T * cos(gim);  // Vertical thrust
    SX Tx_body = T * sin(gim);  // Horizontal thrust
    SX T_body  = SX::vertcat({Tx_body, Tz_body});

    // Rotation matrix
    SX Rot = SX::vertcat({
        SX::horzcat({cos(theta), -sin(theta)}),
        SX::horzcat({sin(theta),  cos(theta)})
    });

    // Thrust in inertial frame
    SX T_inert = mtimes(Rot, T_body);
    SX Tx_inert = T_inert(0);
    SX Tz_inert = T_inert(1);

    // Torque from gimbal
    SX tau = T * sin(gim) * E_z;

    // Dynamics
    SX x_dot      = vx;
    SX z_dot      = vz;
    SX theta_dot  = wy;
    SX vx_dot     = Tx_inert / m;
    SX vz_dot     = Tz_inert / m - g;
    SX wy_dot     = tau / J_y;

    // State derivative vector
    X_dot = SX::vertcat({
        x_dot,
        z_dot,
        theta_dot,
        vx_dot,
        vz_dot,
        wy_dot,
    });

    // CasADi function
    f = Function("f", {X, U}, {X_dot}, {"X", "U"}, {"X_dot"});
}

void Dynamics::setupIntegrator() {

    Function integ;

    SXDict dae = {{"x", X}, {"u", U}, {"p", dt}, {"ode", X_dot*dt}};

    Dict opts;
    if (is_implicit) {
        // Create integrator options
        opts["collocation_scheme"] = "legendre";
        opts["interpolation_order"] = 3;
        opts["simplify"] = true;
        opts["rootfinder"] = "fast_newton";
        
        // Create the integrator function
        integ = integrator("F", "collocation", dae, opts);

    } else {
        //SX X_next = rk4(X_dot, X, dt);
        opts["simplify"] = true;
        integ = integrator("F", "rk", dae, opts);

    }
    // Set the function F to the integrator
    F = integ.map(N_steps, "serial");
}

Function Dynamics::getDiscreteDynamics() const{
    // Returns the integrator function F
    return F;
}

Function Dynamics::getContinuousDynamics() const {
    // Returns the continuous dynamics function f
    return f;
}

int Dynamics::getStateDim() const {
    // Returns the number of states
    return n_X;
}

int Dynamics::getInputDim() const {
    // Returns the number of controls
    return n_U;
}


// RK4 integrator
SX rk4(const SX& x_dot, const SX& x, const SX& dt) {
    SX k1{x_dot};
    SX k2{substitute(x_dot, x, x + dt / 2 * k1)};
    SX k3{substitute(x_dot, x, x + dt / 2 * k2)};
    SX k4{substitute(x_dot, x, x + dt * k3)};
    return x + dt / 6 * (k1 + 2 * k2 + 2 * k3 + k4);
}

File: Python Exercise/src/lib/dynamics/dynamics.py
import casadi as ca
import sys
import os
current_dir = os.path.dirname(os.path.abspath(__file__))
src_dir = os.path.abspath(os.path.join(current_dir, '../..'))
sys.path.append(src_dir)
from simulation_parameters import *

class Dynamics:
    """
    Rocket dynamics class.
    
    Provides both continuous dynamics function f(X, U) -> X_dot
    and discrete integrator F(X, U, dt) -> X_next
    """
    
    def __init__(self, implicit=False, act_dyn=False):
        """Initialize rocket dynamics with physical parameters."""
        
        # Initialize symbolic variables and dynamics
        self.implicit = implicit  # Flag for implicit dynamics
        self.act_dyn = act_dyn    # Flag for actuator dynamics
        self._setup_symbolic_variables()
        if self.act_dyn:
            self._setup_dynamics_with_TVC()
        else:
            self._setup_dynamics()
        self._setup_integrator()
    
    def _setup_symbolic_variables(self):
        """Setup symbolic state and control variables."""
        # States
        self.X = ca.vertcat(
            ca.SX.sym('x'),          # Horizontal position--, m
            ca.SX.sym('z'),          # Vertical position----, m
            ca.SX.sym('theta'),      # Pitch angle----------, rad
            ca.SX.sym('vx'),         # Horizontal velocity--, m/s
            ca.SX.sym('vz'),         # Vertical velocity----, m/s
            ca.SX.sym('wy'),         # Angular velocity-----, rad/s
        )
        if self.act_dyn:
            self.X = ca.vertcat(
                self.X, 
                ca.SX.sym('gim_real'), # Real gimbal angle--, rad
                ca.SX.sym('gim_dot')   # Gimbal angle rate--, rad/s
            )
        self.N_X = self.X.size1()    # Number of states-----, -

        # Controls
        self.U = ca.vertcat(
            ca.SX.sym('T'),          # Engine thrust--------, N
            ca.SX.sym('gimbal'),     # Gimbal angle---------, rad
        )
        self.N_U = self.U.size1()    # Number of controls---, -

        # Time step
        self.dt = ca.SX.sym('dt')    # Time step------------, s
    
    def _setup_dynamics(self):
        """Setup continuous dynamics equations."""
        # Extract state variables
        theta, vx, vz, wy = self.X[2], self.X[3], self.X[4], self.X[5]
        
        # Extract control variables
        T = self.U[0]                # Engine thrust----------------------, N              
        gimbal = self.U[1]           # Gimbal angle-----------------------, rad                 
        
        # Thrust in body frame
        Tz_body = T * ca.cos(gimbal) # Vertical thrust in body frame------, N  
        Tx_body = T * ca.sin(gimbal) # Horizontal thrust in body frame----, N
        T_body = ca.vertcat(Tx_body, Tz_body)  # Total thrust in body frame---------, N

        # Rotation matrix for body-to-inertial frame transformation
        Rot = ca.vertcat(
            ca.horzcat(ca.cos(theta), -ca.sin(theta)), 
            ca.horzcat(ca.sin(theta),  ca.cos(theta))
        )

        # Thrust in inertial frame
        T_inert = ca.mtimes(Rot, T_body)       # Total thrust in inertial frame-----, N
        Tx_inert = T_inert[0]                  # Horizontal thrust in inertial frame, N
        Tz_inert = T_inert[1]                  # Vertical thrust in inertial frame--, N

        # Torque
        tau = T * ca.sin(gimbal) * E_z         # Torque due to gimbal---------------, Nm

        # State derivatives
        x_dot = vx                             # Horizontal velocity----------------, m/s 
        z_dot = vz                             # Vertical velocity------------------, m/s   
        theta_dot = wy                         # Angular velocity-------------------, rad/s  
        vx_dot = Tx_inert / m                  # Horizontal acceleration------------, m/s^2
        vz_dot = Tz_inert / m - g              # Vertical acceleration--------------, m/s^2
        wy_dot = tau / J_y                     # Angular acceleration---------------, rad/s^2

        # Complete state derivative vector
        self.X_dot = ca.vertcat(x_dot, z_dot, theta_dot, vx_dot, vz_dot, wy_dot)

        # Create continuous dynamics function
        self.f = ca.Function('f', [self.X, self.U], [self.X_dot], 
                            ["X", "U"], ["X_dot"])
        
    def _setup_dynamics_with_TVC(self):
        """Setup continuous dynamics equations."""
        # Extract state variables
        theta, vx, vz, wy = self.X[2], self.X[3], self.X[4], self.X[5]
        gim_real, gim_dot = self.X[6], self.X[7]
        # Extract control variables
        T = self.U[0]                # Engine thrust----------------------, N              
        gimbal = self.U[1]           # Gimbal angle-----------------------, rad                 
        
        # Thrust in body frame
        Tz_body = T * ca.cos(gim_real) # Vertical thrust in body frame------, N  
        Tx_body = T * ca.sin(gim_real) # Horizontal thrust in body frame----, N
        T_body = ca.vertcat(Tx_body, Tz_body)  # Total thrust in body frame---------, N

        # Rotation matrix for body-to-inertial frame transformation
        Rot = ca.vertcat(
            ca.horzcat(ca.cos(theta), -ca.sin(theta)), 
            ca.horzcat(ca.sin(theta),  ca.cos(theta))
        )

        # Thrust in inertial frame
        T_inert = ca.mtimes(Rot, T_body)       # Total thrust in inertial frame-----, N
        Tx_inert = T_inert[0]                  # Horizontal thrust in inertial frame, N
        Tz_inert = T_inert[1]                  # Vertical thrust in inertial frame--, N

        # Torque
        tau = T * ca.sin(gim_real) * E_z         # Torque due to gimbal---------------, Nm

        # State derivatives
        x_dot = vx                             # Horizontal velocity----------------, m/s 
        z_dot = vz                             # Vertical velocity------------------, m/s   
        theta_dot = wy                         # Angular velocity-------------------, rad/s  
        vx_dot = Tx_inert / m                  # Horizontal acceleration------------, m/s^2
        vz_dot = Tz_inert / m - g              # Vertical acceleration--------------, m/s^2
        wy_dot = tau / J_y                     # Angular acceleration---------------, rad/s^2
        gim_ddot = -2 * zeta * w_n * gim_dot - w_n**2 * gim_real + w_n**2 * gimbal

        # Complete state derivative vector
        self.X_dot = ca.vertcat(x_dot, z_dot, theta_dot, vx_dot, vz_dot, wy_dot, gim_dot, gim_ddot)

        # Create continuous dynamics function
        self.f = ca.Function('f', [self.X, self.U], [self.X_dot], 
                            ["X", "U"], ["X_dot"])
    
    def _setup_integrator(self):
        """Setup discrete integrator using collocation method."""
        # Define DAE for integrator
        if self.implicit:
            dae = {
                'x': self.X,
                'u': self.U,
                'p': self.dt,
                'ode': self.X_dot * self.dt
            }
            
            # Create integrator
            self.F = ca.integrator('F', 'collocation', dae)
        else:
            # Explicit integrator for non-implicit case
            X_next = self._rk4_explicit(self.f, self.X, self.U, self.dt)
            self.F = ca.Function('F', [self.X, self.U, self.dt], [X_next],
                                ["X", "U", "dt"], ["X_next"])
    
    def _rk4_explicit(self, f, x, u, h):
        """ One step of ode integration with Runge-Kutta of order 4

        Args:
            f:  implementation of an ode, such that dx/dt = f(x, u)
            x:  initial value
            u:  control input
            h: time step
        """
        k1       = f(x, u)
        k2       = f(x + h/2 * k1, u)
        k3       = f(x + h/2 * k2, u)
        k4       = f(x + h * k3, u)
        return x + h/6 * (k1 + 2*k2 + 2*k3 + k4)


    def get_continuous_dynamics(self):
        """
        Get continuous dynamics function.
        
        Returns:
            ca.Function: f(X, U) -> X_dot
        """
        return self.f
    
    def get_discrete_integrator(self):
        """
        Get discrete integrator function.
        
        Returns:
            ca.Function: F(X, U, dt) -> X_next
        """
        return self.F
    
    def get_state_dim(self):
        """Get state dimension."""
        return self.N_X
    
    def get_control_dim(self):
        """Get control dimension."""
        return self.N_U
    
    def is_implicit(self):
        """Check if the dynamics are implicit.
        Returns:
            bool: True if implicit, False otherwise
        """
        return self.implicit


if __name__ == "__main__":
    # Create rocket dynamics instance
    rocket = Dynamics()
    
    # Get dynamics functions
    f = rocket.get_continuous_dynamics()
    F = rocket.get_discrete_integrator()
    
    print(f"Continuous dynamics function: {f}")
    print(f"Discrete integrator: {F}")